### PR TITLE
Add retry logic to S3 uploads

### PR DESCRIPTION
**Pull request by alexgandy**
Also:
1. Update node version
2. Add CircleCI
3. Use dotenv instead of a credentials file.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Javascript Node CircleCI 2.0 configuration file
+#
+# Check https://circleci.com/docs/2.0/language-javascript/ for more details
+#
+version: 2
```
Comment (cdeveas) [inline .circleci/config.yml]: woo

File: lib/assetProcessor.js [hunk 2]
```diff
@@ -381,15 +381,15 @@ AssetProcessor.prototype.uploadExtrasToCdn = function(callback) {
...
+    const imports = self.config.targets && self.config.targets.stylesheets && self.config.targets.stylesheets.imports || {};
 
-    var destDir = path.resolve(self.config.stylesheetsRoot, imports.destination || '');
-    var sources = imports.sources || [];
-    var mappings = imports.mappings || {};
-    var numImported = 0;
+    const destDir = path.resolve(self.config.stylesheetsRoot, imports.destination || '');
+    const sources = imports.sources || [];
+    const mappings = imports.mappings || {};
+    let numImported = 0;
```
Comment (cdeveas) [inline lib/assetProcessor.js]: ES6 boy scout badge

File: lib/assetProcessor.js [hunk 3]
```diff
@@ -413,6 +413,8 @@ AssetProcessor.prototype.importLatestStylesheets = function(callback) {
                     eachNext(err);
                 });
             } else {
+                console.
+                console.error(err);
```
Comment (cdeveas) [inline lib/assetProcessor.js]: Typo?

File: lib/s3.js [hunk 4]
```diff
@@ -51,8 +51,8 @@ S3.prototype.getBucket = function() {
  * Uploads the given text contents to S3
  * @param {string|Buffer} contents the text or buffer to upload
  * @param {string} path the path the contents should be saved to
- * @param {bool|object} makePrivateOrHeaders whether or not to make the file private on S3; or the full set of heders to use.
- * @param {function} callback function(err, url)
+ * @param {bool|object|function} makePrivateOrHeaders whether or not to make the file private on S3; or the full set of heders to use.
```
Comment (cdeveas) [inline lib/s3.js]: If this is an optional var I generally do:
```javascript
* @param {bool|object} [makePrivateOrHeaders = false] - Optional argument which...
```

I _think_ that's right according to http://usejsdoc.org/? There's not a clear way to do most things in jsdoc land.

File: package.json [hunk 5]
```diff
@@ -3,12 +3,15 @@
   "version": "0.1.3",
   "scripts": {
     "start": "node process.js",
-    "test": "nodeunit"
+    "test": "node ./node_modules/.bin/nodeunit"
```
Comment (cdeveas) [inline package.json]: ./node_modules/.bin/nodeunit isn't an executable?

File: lib/assetProcessor.js (hunk 2 above)
Comment (alexgandy) [reply lib/assetProcessor.js]: Thumbs Up

File: package.json (hunk 5 above)
Comment (alexgandy) [reply package.json]: Hey, it is. And it works with the node caller too. Cool. Fixed.

File: lib/s3.js (hunk 4 above)
Comment (alexgandy) [reply lib/s3.js]: I'm actually making it a required parameter. Every usage of it passes a callback, so I think this was a case of @Zugwalt being too nice and over-engineering a bit.

File: lib/s3.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
 'use strict';
 
-var fs      = require('fs');
-var path    = require('path');
-
-var knox    = require('knox');
-var mime    = require('mime');
+const fs      = require('fs');
+const path    = require('path');
+const knox    = require('knox');
```
Comment (Zugwalt) [inline lib/s3.js]: Shit we are still using `knox`? We really shouldn't and should use their SDK.

File: lib/s3.js [hunk 7]
```diff
@@ -63,15 +63,34 @@ S3.prototype.putBuffer = function(contents, path, makePrivateOrHeaders, callback
...
-    var buffer = contents instanceof Buffer ? contents : new Buffer(contents);
+    const self = this;
+    const buffer = contents instanceof Buffer ? contents : new Buffer(contents);
+
+    const maxTries = 3;
+    let numTries = 0;
+    let fileUpload = null;
+
+    const untilEnd = () => { return fileUpload || numTries > maxTries };
+    const untilIterator = (untilNext) => {
```
Comment (Zugwalt) [inline lib/s3.js]: I find it easier to read when you just inline this in the `async()` code. Or at a minimum declare later in the file. Clean code a good assertion that the code should be like a newspaper article--high level at first and later on more details.

File: lib/s3.js [hunk 8]
```diff
@@ -63,15 +63,34 @@ S3.prototype.putBuffer = function(contents, path, makePrivateOrHeaders, callback
...
+        self.client.putBuffer(buffer, path, _createHeaders(makePrivateOrHeaders), function (err, result) {
```
Comment (Zugwalt) [inline lib/s3.js]: why `=>` above and `function()` here? Don't think you need access to `this`.

File: lib/s3.js [hunk 9]
```diff
@@ -63,15 +63,34 @@ S3.prototype.putBuffer = function(contents, path, makePrivateOrHeaders, callback
...
+
+            fileUpload = result && self.urlWithBucket(path);
+            let delayTimeoutMs = 0;
 
-    self.client.putBuffer(buffer, path, _createHeaders(makePrivateOrHeaders), function(err, result) {
-        if (!err && (!result || result.statusCode !== 200)) {
-            err = new Error('Status code other 200 from AWS: '+result.statusCode);
-        }
+            if (++numTries <= maxTries && err) {
+                delayTimeoutMs = Math.pow(2, 10+numTries); //exponential backoff before retrying
```
Comment (Zugwalt) [inline lib/s3.js]: Nice with exponential backoff!

File: lib/s3.js [hunk 10]
```diff
@@ -63,15 +63,34 @@ S3.prototype.putBuffer = function(contents, path, makePrivateOrHeaders, callback
...
+                console.log('S3 Upload error, will retry', {err, numTries, delayTimeoutMs});
```
Comment (Zugwalt) [inline lib/s3.js]: Minor but I think this is better suited for `console.warn()`

File: lib/s3.js [hunk 11]
```diff
@@ -85,24 +104,40 @@ S3.prototype.putBuffer = function(contents, path, makePrivateOrHeaders, callback
...
     headers['Content-Length'] = fs.statSync(sourcePath).size;
 
-    var fileUpload = self.client.putStream(stream, targetPath, headers, function(err, result){
-        callback(err, result && self.urlWithBucket(targetPath));
-    });
+    const maxTries = 3;
+    let numTries = 0;
+    let fileUpload = null;
+
+    const untilEnd = () => { return fileUpload || numTries > maxTries };
```
Comment (Zugwalt) [inline lib/s3.js]: This retry boilerplate should be extracted into its own function instead of copy/pasted.

File: .gitignore [hunk 12]
```diff
@@ -2,7 +2,7 @@
 node_modules
 
 #credentials used by unit tests
-test/credentials.json
+.env
```
Comment (Zugwalt) [inline .gitignore]: Why get rid of `credentials.json`?

File: test/test-assetProcessor.js [hunk 13]
```diff
@@ -1,41 +1,36 @@
...
-            "secret": "aws-s3-secret"
-        },
-        "git": {
-            "token": "github-token"
-        }
-    }, undefined, 2));
+const th = new NodeunitAsync();
+
+if (_.any(['S3_BUCKET', 'S3_ACCESS_KEY', 'S3_SECRET', 'GITHUB_OAUTH_TOKEN'], k => { return _.isUndefined(process.env[k]); })) {
+    throw new Error("Missing required configuration. Do you have a .env file?");
```
Comment (Zugwalt) [inline test/test-assetProcessor.js]: Why replace the more helpful error message with this?

File: .gitignore (hunk 12 above)
Comment (alexgandy) [reply .gitignore]: It was only used for the tests. And I wanted to get this added to circleci...

File: lib/s3.js (hunk 9 above)
Comment (alexgandy) [reply lib/s3.js]: Haha, fairly sure you wrote it in wasabi...which is where I stole this all from.

File: lib/s3.js (hunk 10 above)
Comment (alexgandy) [reply lib/s3.js]: Good call, will update.

File: lib/s3.js (hunk 11 above)
Comment (alexgandy) [reply lib/s3.js]: I agree. I want to do a deeper dive at cleaning up all of this code...

File: .gitignore (hunk 12 above)
Comment (alexgandy) [reply .gitignore]: And since this is a public repo, we can commit API access keys anywhere...

File: test/test-assetProcessor.js (hunk 13 above)
Comment (alexgandy) [reply test/test-assetProcessor.js]: There's already an example dotenv file. And we aren't using credentials.json anywhere else.

Comment (Zugwalt) [reply test/test-assetProcessor.js]: Ok you've won me over with the dotenv

File: lib/s3.js (hunk 7 above)
Comment (alexgandy) [reply lib/s3.js]: Fixed.

File: lib/s3.js (hunk 8 above)
Comment (alexgandy) [reply lib/s3.js]: Honestly, just force of habit I think. Fixed.

**Pull request merged**